daemon/media_socket.c

```diff
@@ -226,14 +224,140 @@ static GHashTable *__logical_intf_name_family_hash;
 static GHashTable *__intf_spec_addr_type_hash;
 static GQueue __preferred_lists_for_family[__SF_LAST];
 
+static unsigned int selection_index = 0;
+static unsigned int selection_count = 0;
+
+
+/* checks for free no_ports on a local interface */
+static int has_free_ports_loc(struct local_intf *loc, unsigned int num_ports) {
+	if (loc == NULL) {
+		ilog(LOG_ERR, "has_free_ports_loc - NULL local interface");
+		return 0;
+	}
+	
+	if (num_ports > loc->spec->port_pool.free_ports) {
+		ilog(LOG_ERR, "Didn't found %d ports available for %.*s/%s",
+			num_ports, loc->logical->name.len, loc->logical->name.s,
+			sockaddr_print_buf(&loc->spec->address.addr));
+		return 0;
+	}
+
+	__C_DBG("Found %d ports available for %.*s/%s from total of %d free ports",
+		num_ports, loc->logical->name.len, loc->logical->name.s,
+		sockaddr_print_buf(&loc->spec->address.addr),
+		loc->spec->port_pool.free_ports);
+
+	return 1;
+}
+
+/* checks for free num_ports on at least one local interface of a logical interface */
+static int has_free_ports_log_any(struct logical_intf *log, unsigned int num_ports) {
+	if (log == NULL) {
+		ilog(LOG_ERR, "has_free_ports_log_any - NULL logical interface");
+		return 0;
+	}
+
+	struct local_intf *loc;
+	GList *l;
+
+	for (l = log->list.head; l; l = l->next) {
+		loc = l->data;
+
+		if (has_free_ports_loc(loc, num_ports)) {
+			return 1;
+		}
+	}
+
+	return 0;
+}
+
+/* checks for free num_ports on all local interfaces of a logical interface */
+static int has_free_ports_log_all(struct logical_intf *log, unsigned int num_ports) {
+	if (log == NULL) {
+		ilog(LOG_ERR, "has_free_ports_log_all - NULL logical interface");
+		return 0;
+	}
+
+	struct local_intf *loc;
+	GList *l;
+
+	for (l = log->list.head; l; l = l->next) {
+		loc = l->data;
+
+		if (!has_free_ports_loc(loc, num_ports)) {
+			return 0;
+		}
+	}
+
+	return 1;
+}
+
+/* run round-robin-calls algorithm */ 
+static struct logical_intf* run_round_robin_calls(GQueue *q, unsigned int num_ports) {
+	struct logical_intf *log = NULL;
+	unsigned int index = 0;
+	unsigned int nr_tries = 0;
+	unsigned int nr_logs = 0;
+	GList *log_iter;
+
+
+	nr_logs = g_queue_get_length(q);
+	log_iter = q->head;
 
+select_log:
+	// choose the next logical interface
+	index = 0;
+	while (index < selection_index && log_iter) {
```

Comment:
This can be shortened a bit by using g_queue_peek_nth() or g_queue_peek_nth_link()